File: src/main/java/dev/entree/vchest/JDBCContext.java
package dev.entree.vchest;

import lombok.With;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.util.logging.Logger;

@With
public record JDBCContext(Logger logger, String host, String port, String username, String password, String dbName,
                          String protocol, @Nullable File dir, String className, @Nullable ClassLoader classLoader) {
    public JDBCContext(Logger logger, String host, String port, String username, String password, String dbName, String protocol, @Nullable File dir, String className, ClassLoader classLoader) {
        this.logger = logger;
        this.host = host;
        this.port = port;
        this.username = username;
        this.password = password;
        this.dbName = dbName;
        this.protocol = protocol;
        this.dir = dir;
        this.className = className;
        this.classLoader = classLoader;
    }

    public static JDBCContext ofMariaDB(Logger logger, String host, String port, String username, String password, String dbName) {
        return new JDBCContext(logger, host, port, username, password, dbName, "mariadb", null, "org.mariadb.jdbc.Driver", null);
    }

    public static JDBCContext ofMySQL(Logger logger, String host, String port, String username, String password, String dbName) {
        return new JDBCContext(logger, host, port, username, password, dbName, "mysql", null, "com.mysql.cj.jdbc.Driver", null);
    }

    public static JDBCContext ofSqlite(Logger logger, String dbName, File dir) {
        return new JDBCContext(logger, "", "", "", "", dbName, "sqlite", dir, "", null);
    }
}

File: src/main/kotlin/dev/entree/vchest/JDBCContext.kt
package dev.entree.vchest

import java.io.File
import java.util.logging.Logger

data class JDBCContext(
    var logger: Logger,
    var host: String,
    var port: String,
    var username: String,
    var password: String,
    var dbName: String,
    var protocol: String,
    var dir: File?,
    var className: String,
    var classLoader: ClassLoader?,
) {
    companion object {
        fun ofMariaDB(
            logger: Logger,
            host: String,
            port: String,
            username: String,
            password: String,
            dbName: String,
        ): JDBCContext {
            return JDBCContext(
                logger,
                host,
                port,
                username,
                password,
                dbName,
                "mariadb",
                null,
                "org.mariadb.jdbc.Driver",
                null
            )
        }

        fun ofMySQL(
            logger: Logger,
            host: String,
            port: String,
            username: String,
            password: String,
            dbName: String,
        ): JDBCContext {
            return JDBCContext(
                logger,
                host,
                port,
                username,
                password,
                dbName,
                "mysql",
                null,
                "com.mysql.cj.jdbc.Driver",
                null
            )
        }

        fun ofSqlite(logger: Logger, dbName: String, dir: File): JDBCContext {
            return JDBCContext(logger, "", "", "", "", dbName, "sqlite", dir, "", null)
        }
    }
}